src/compile/data/optimizers.ts

```diff
@@ -78,3 +78,29 @@ export function removeUnusedSubtrees(node: DataFlowNode) {
   }
   return true;
 }
+
+/**
+ * Generates time unit nodes from selection filters and removes duplicate
+ * time unit nodes (as determined by the name of the output field)
+ * that may be generated due to selections projected over time units.
+ */
+export function cleanupSelectionTimeUnits(leaf: DataFlowNode) {
+  let fields = {};
+  return iterateFromLeaves((node: DataFlowNode) => {
+    if (node instanceof FilterNode) {
+      // Selection filters only add TimeUnitNodes on assembly.
```

Comment:
Yeah, I agree -- adding the node during assembly doesn't feel great. Definitely open to better implementation ideas. Here're the constraints I was working within, which should hopefully answer your questions:

1. This PR is necessary for selections projected over time units that are applied in a _different_ view from that which they are defined in. As a result, while _constructing_ the necessary `TimeUnitNode` is feasible while parsing a selection, _inserting_ it into the necessary dataflows isn't during the selection parsing _or_ assembly process (selection assembly again only occurs in the views it is defined within). 
2. Selection application is centralized in the `predicate` function in `selection.ts`. It is called by both `mixins` (to generate Vega production rules) and filter's `expression`. 
3. Parsing data transforms occurs before parsing selections.